be/src/exec/pipeline/pipeline_driver_queue.h

```diff
@@ -49,7 +70,50 @@ class SubQuerySharedDriverQueue {
 
     double accu_time_after_divisor() { return _accu_consume_time.load() / factor_for_normal; }
 
-    std::queue<DriverRawPtr> queue;
+    void put(const DriverRawPtr driver) {
+        if (driver->driver_state() == DriverState::CANCELED) {
+            queue.emplace_front(driver);
+        } else {
+            queue.emplace_back(driver);
+        }
+        driver_number++;
+    }
+
+    void cancel(const DriverRawPtr driver) { pending_cancel_queue.emplace(driver); }
+
+    DriverRawPtr take() {
+        DCHECK(!empty());
+        if (!pending_cancel_queue.empty()) {
+            DriverRawPtr driver = pending_cancel_queue.front();
+            pending_cancel_queue.pop();
+            cancelled_set.insert(driver);
+            --driver_number;
+            return driver;
+        }
+
+        while (!queue.empty()) {
+            DriverRawPtr driver = queue.front();
+            queue.pop_front();
+            auto iter = cancelled_set.find(driver);
+            if (iter != cancelled_set.end()) {
+                cancelled_set.erase(iter);
```

Comment:
has Driver::finalize function been invoked before erase driver in the cancelled queue?

Reply:
> has Driver::finalize function been invoked before erase driver in the cancelled queue?

yes, if its raw ptr is in cancelled set, it means this Driver has been taken out from DriverQueue